include/openPMD/IO/ADIOS/ADIOS2IOHandler.hpp

```diff
@@ -21,64 +21,595 @@
+     */
+    template < typename, typename = void > struct DatasetHelper;
+
+    template < typename T >
+    struct DatasetHelper<
+        T, typename std::enable_if< DatasetTypes< T >::validType >::type >
+    {
+        openPMD::ADIOS2IOHandlerImpl * m_impl;
+
+
+        explicit DatasetHelper( openPMD::ADIOS2IOHandlerImpl * impl );
+
+
+        void openDataset( InvalidatableFile, const std::string & varName,
+                          Parameter< Operation::OPEN_DATASET > & parameters );
+
+        void readDataset( BufferedGet &, adios2::IO &, adios2::Engine &,
+                          std::string const & fileName );
+
+        static void
+        defineVariable( adios2::IO & IO, const std::string & name,
+                        std::unique_ptr< adios2::Operator > compression,
+                        const adios2::Dims & shape, const adios2::Dims & start,
+                        const adios2::Dims & count, bool constantDims );
+
+        void writeDataset( BufferedPut &, adios2::IO &, adios2::Engine & );
+    };
+
+    template < typename T >
+    struct DatasetHelper<
+        T, typename std::enable_if< !DatasetTypes< T >::validType >::type >
+    {
+        explicit DatasetHelper( openPMD::ADIOS2IOHandlerImpl * impl );
+
+
+        static void throwErr( );
+
+        template < typename... Params > void openDataset( Params &&... );
+
+        template < typename... Params > void readDataset( Params &&... );
+
+        template < typename... Params >
+        static void defineVariable( Params &&... );
+
+        template < typename... Params > void writeDataset( Params &&... );
+    };
+
+    // Other datatypes used in the ADIOS2IOHandler implementation
+
+
+    struct BufferedActions;
+
+    /*
+     * IO-heavy action to be executed upon flushing.
+     */
+    struct BufferedAction
+    {
+        virtual ~BufferedAction( ) = default;
+
+        virtual void run( BufferedActions & ) = 0;
+    };
+
+    struct BufferedGet : BufferedAction
+    {
+        std::string name;
+        Parameter< Operation::READ_DATASET > param;
+
+        void run( BufferedActions & ) override;
+    };
+
+    struct BufferedPut : BufferedAction
+    {
+        std::string name;
+        Parameter< Operation::WRITE_DATASET > param;
+
+        void run( BufferedActions & ) override;
+    };
+
+    struct BufferedAttributeRead : BufferedAction
+    {
+        Parameter< Operation::READ_ATT > param;
+        std::string name;
+
+        void run( BufferedActions & ) override;
+    };
+
+    /*
+     * Manages per-file information about
+     * (1) the file's IO and Engine objects
+     * (2) the file's deferred IO-heavy actions
+     */
+    struct BufferedActions
+    {
+        BufferedActions( BufferedActions const & ) = delete;
+
+        std::string m_file;
+        adios2::IO m_IO;
+        std::vector< std::unique_ptr< BufferedAction > > m_buffer;
+        // std::optional would be more idiomatic, but it's not in
```

Comment:
feel free to make this a doxygen string and add a `@todo ...` here, so we can refactor it when we switch at some point to C++17...